File: src/lib/progress.ts
import { listen, type UnlistenFn } from '@tauri-apps/api/event';

export const EVT = {
  SettingsChanged: 'settings://changed',
  ToolInstallProgress: 'tool://install-progress',
  TaskProgress: 'task://progress',
  TaskLog: 'task://log',
  TaskDone: 'task://done',
  TaskError: 'task://error',
} as const;

export type TaskLogLevel = 'info' | 'warn' | 'error';

export type ProgressEvent = {
  task_id: string;
  percent: number;
  stage?: string;
};

export type LogEvent = {
  task_id: string;
  line: string;
  level: TaskLogLevel;
};

export type DoneEvent = {
  task_id: string;
  result?: unknown;
};

export type ErrorEvent = {
  task_id: string;
  error: string;
};

export function onTaskProgress(cb: (e: ProgressEvent) => void): Promise<UnlistenFn> {
  return listen<ProgressEvent>(EVT.TaskProgress, (ev) => cb(ev.payload));
}

export function onTaskLog(cb: (e: LogEvent) => void): Promise<UnlistenFn> {
  return listen<LogEvent>(EVT.TaskLog, (ev) => cb(ev.payload));
}

export function onTaskDone(cb: (e: DoneEvent) => void): Promise<UnlistenFn> {
  return listen<DoneEvent>(EVT.TaskDone, (ev) => cb(ev.payload));
}

export function onTaskError(cb: (e: ErrorEvent) => void): Promise<UnlistenFn> {
  return listen<ErrorEvent>(EVT.TaskError, (ev) => cb(ev.payload));
}

File: src-tauri/src/progress.rs
use serde::Serialize;
use tauri::{AppHandle, Emitter};

pub fn emit_progress(app: &AppHandle, task_id: &str, percent: f32, stage: &str) {
    let _ = app.emit(
        "task://progress",
        serde_json::json!({ "task_id": task_id, "percent": percent.clamp(0.0, 100.0), "stage": stage }),
    );
}

pub fn emit_log(app: &AppHandle, task_id: &str, line: &str, level: &str) {
    let _ = app.emit(
        "task://log",
        serde_json::json!({ "task_id": task_id, "line": line, "level": level }),
    );
}

pub fn emit_done<V: Serialize>(app: &AppHandle, task_id: &str, result: Option<V>) {
    let payload = match result {
        Some(r) => serde_json::json!({ "task_id": task_id, "result": r }),
        None => serde_json::json!({ "task_id": task_id }),
    };
    let _ = app.emit("task://done", payload);
}

pub fn emit_error(app: &AppHandle, task_id: &str, error: &str) {
    let _ = app.emit(
        "task://error",
        serde_json::json!({ "task_id": task_id, "error": error }),
    );
}
